### ISSUE TITLE
Lock constraint always aligns rotation of the 2 bodies?

### ISSUE THREAD

**Issue by josephrocca** [schteppe/cannon.js#225]
Has anyone else ever had this problem? I've been trying to nut it out all day.
![lock constraint](https://cloud.githubusercontent.com/assets/1167575/9288903/65f65264-4390-11e5-8ecf-6c98a34c2197.png)
I initialise the two boxes like the picture on the left, but when I apply the `LockConstraint`, both boxes quickly rotate around so that they are aligned (but offset) like the picture on the right, and then lock into place like that.

My code for applying the constraint is simply:

```
      var c = new CANNON.LockConstraint(bodyA, bodyB, {
         maxForce: params.maxForce || Number.MAX_VALUE
      });
      c.collideConnected = false;
      this.world.addConstraint(c);
```

I haven't fully tested other constraint types for similar behavior, but preliminary tests show similar behavior - if the bodies have different orientations, they will be aligned when I add the constraint.

**Comment by schteppe** [comment]
Are you using the very latest version? I think this was fixed in https://github.com/schteppe/cannon.js/commit/20b7eee77c1fc6c97426c112818a94fef926461d

Rebuild and try again? Would be nice to have this fix verified. Thanks.

**Comment by josephrocca** [comment]
Awesome, so I can confirm that `LockConstraint` is now working properly after updating to the latest version, but `HingeConstraint` is still having the exact same problem :\

**Comment by schteppe** [comment]
Oh, nice. The HingeConstraint doesn't have that feature yet. But I guess it would be very handy :)

**Comment by josephrocca** [comment]
Hmm, so the initial orientation of the 2 bodies must be aligned for a hinge constraint to work properly? If that's correct, are there any hacky workarounds for this that you could suggest?

**Comment by schteppe** [comment]
No, you just need to specify the pivots and the axes correctly...

As you probably know, the hinge constraint needs a local axis and pivot point in each body. The constraint tries to match the pivot point positions, and the axis directions.

The trick is to setup the parameters so the constraint is in rest when the simulation starts. This can be done manually or in a more automatic manner. The ```LockConstraint``` is fully automatic and need no parameters at all - it will take the positions and orientations of the bodies as initial parameters and compute its local pivots and axes.

The HingeConstraint cannot be made as convenient as the LockConstraint, but almost. If you know the point and axis (in world space) of your hinge, you can create it like this:

```js
// Beware: untested code
var worldPoint = ...
var worldAxis = ...
var localPivotA = bodyA.pointToLocalFrame(worldPoint);
var localPivotB = bodyB.pointToLocalFrame(worldPoint);
var localAxisA = bodyA.vectorToLocalFrame(worldAxis);
var localAxisB = bodyB.vectorToLocalFrame(worldAxis);
var c = new CANNON.HingeConstraint(bodyA, bodyB, {
    pivotA: localPivotA,
    pivotB: localPivotB,
    axisA: localAxisA,
    axisB: localAxisB
});
```

If you use this code, the constraint should start at rest, and not get an "alignment movement" at start.

Should be really easy to get this into the constructor of ```HingeConstraint```, I've just not gotten there yet...

**Comment by josephrocca** [comment]
Ahhh, yes, you're right! I wasn't setting my axes right :| thanks for your patience!

**Comment by donmccurdy** [comment]
@schteppe Would you mind doing a release with this patch, if it's not blocked on a milestone? I'm using it in a [Vive example](https://github.com/donmccurdy/aframe-extras/commit/99f9f5d7d0f35f1cc53ebcbc01f63500899501a6#diff-a6d399e94d66267f2329f56b87fa81d8R68), and don't want anyone reusing my code to be surprised by that.

**Comment by jsoutherland** [comment]
@schteppe I noticed the most recent release (0.6.2) is still missing this patch.
I replaced my version with master (569730f94a1d9da47967a24fad0323ef7d5b4119)  and believe I am still seeing this issue with lockConstraint.

I have tried turning off self collisions and tried removing one body's collision shape and neither worked.

**Comment by schteppe** [comment]
Yeah, should really make some time for a new release..
Did you build the latest master? The build/ directory is only updated on releases. Run grunt to update it manually

**Comment by jsoutherland** [comment]
@schteppe Thanks I missed that step.  

I tried again but it still didn't work. I may have something else going on.

I did have to use this fix for grunt to succeed:  https://github.com/schteppe/cannon.js/pull/280
I was working off of https://github.com/schteppe/cannon.js/pull/275

**Comment by jsoutherland** [comment]
@schteppe I don't see how the three rotation equations are different?

https://github.com/schteppe/cannon.js/blob/20b7eee77c1fc6c97426c112818a94fef926461d/src/constraints/LockConstraint.js#L45-L62

**Comment by schteppe** [comment]
The equations are the same but the vectors they operate on are different. See the diff here: https://github.com/schteppe/cannon.js/commit/20b7eee77c1fc6c97426c112818a94fef926461d

**Comment by jsoutherland** [comment]
Thank you for pointing that out.  I see how it works now.

It is working now for me, Chrome was caching the old version :disappointed:

**Comment by deltakosh** [comment]
Hello, we have this issue with the latest version we found on the repo. Is this fix available somewhere? Do you think you could build a new release?

**Comment by donmccurdy** [comment]
Physics for A-Frame VR currently uses https://github.com/donmccurdy/cannon.js/tree/v0.6.2-dev1, to include this fix.

**Comment by deltakosh** [comment]
Thanks @donmccurdy, but I would have expected an official build. Some users would like to get the library from the official repo :(

**Comment by roscopetracula** [comment]
@schteppe Bump on the new build -- Babylon.js could benefit from that too -- the standard method of loading Cannon there is to pull it in via cdn, so building is beyond what many user are likely to know to do.

Thank you for the great library!! :-)
